File: orchestrator/tools/bash.py
"""Bash tool implementation."""

import os
import subprocess
import time
from typing import TYPE_CHECKING, Any, Dict, Optional

from rich.live import Live
from rich.text import Text

from ..display import ICONS, AnimatedWaiter
from ..display_adapter import get_display
from .base import BaseTool, ToolResult

if TYPE_CHECKING:
    from ..context import ExecutionContext
    from ..tmux import TmuxManager


class BashTool(BaseTool):
    """Execute bash commands in subprocess or tmux pane."""

    @property
    def name(self) -> str:
        return "bash"

    def validate_step(self, step: Dict[str, Any]) -> None:
        """Validate bash step configuration."""
        if "command" not in step:
            raise ValueError("Bash step requires 'command' field")

    def execute(
        self,
        step: Dict[str, Any],
        context: "ExecutionContext",
        tmux_manager: "TmuxManager",
    ) -> ToolResult:
        """Execute bash command.

        If visible=True, runs in tmux pane.
        If visible=False (default), runs in background subprocess.

        Supports 'env' option to pass variables as environment variables,
        which safely handles shell-breaking characters in variable values.
        """
        command = context.interpolate(step["command"])
        # Always default to project_path for cwd
        cwd = context.interpolate_optional(step.get("cwd")) or str(context.project_path)
        visible = step.get("visible", False)
        strip_output = step.get("strip_output", True)

        # Build environment variables if specified
        env = self._build_env(step.get("env"), context)

        if visible:
            return self._execute_visible(command, cwd, tmux_manager, strip_output, env)
        else:
            return self._execute_subprocess(command, cwd, strip_output, env)

    def _build_env(
        self,
        env_config: Optional[Dict[str, str]],
        context: "ExecutionContext",
    ) -> Optional[Dict[str, str]]:
        """Build environment variables dict from step config.

        Args:
            env_config: Dict of env var names to values (may contain {var} placeholders)
            context: Execution context for interpolation

        Returns:
            Combined environment (system + custom) or None if no custom env specified
        """
        if env_config is None:
            return None

        # Start with copy of current environment
        env = os.environ.copy()

        # Add/override with custom variables (interpolated)
        for key, value in env_config.items():
            env[key] = context.interpolate(str(value))

        return env

    def _execute_subprocess(
        self,
        command: str,
        cwd: str | None,
        strip_output: bool,
        env: Optional[Dict[str, str]] = None,
    ) -> ToolResult:
        """Execute command in background subprocess.

        Args:
            command: Shell command to execute
            cwd: Working directory for command
            strip_output: Whether to strip whitespace from output
            env: Environment variables (None = inherit system env)
        """
        display = get_display()
        display.print_bash_running(command)

        try:
            process = subprocess.run(
                command,
                shell=True,
                cwd=cwd,
                capture_output=True,
                text=True,
                timeout=600,  # 10 minute timeout
                env=env,
            )

            output = process.stdout or ""
            if process.stderr:
                output += f"\n[STDERR]\n{process.stderr}"

            if strip_output:
                output = output.strip()

            success = process.returncode == 0

            return ToolResult(
                success=success,
                output=output,
                error=process.stderr if not success else None,
            )
        except subprocess.TimeoutExpired:
            return ToolResult(
                success=False,
                error="Command timed out after 10 minutes",
            )
        except Exception as e:
            return ToolResult(
                success=False,
                error=str(e),
            )

    def _execute_visible(
        self,
        command: str,
        cwd: str | None,
        tmux_manager: "TmuxManager",
        strip_output: bool,
        env: Optional[Dict[str, str]] = None,
    ) -> ToolResult:
        """Execute command in visible tmux pane.

        Args:
            command: Shell command to execute
            cwd: Working directory for command
            tmux_manager: Tmux pane manager
            strip_output: Whether to strip whitespace from output
            env: Environment variables (currently not supported in visible mode)
        """
        # For visible mode with custom env vars, wrap the command with exports
        if env:
            # Build export statements for custom env vars
            exports = []
            for key, value in env.items():
                # Skip system env vars that are already set
                if key not in os.environ or os.environ[key] != value:
                    # Escape single quotes in value for shell safety
                    escaped_value = value.replace("'", "'\\''")
                    exports.append(f"export {key}='{escaped_value}'")
            if exports:
                command = " && ".join(exports) + " && " + command

        # Launch bash pane
        tmux_manager.launch_bash_pane(command, cwd)

        try:
            # Wait for completion using idle detection
            output = self._wait_for_completion(tmux_manager)

            if strip_output:
                output = output.strip()

            return ToolResult(
                success=True,  # Can't easily determine exit code in tmux
                output=output,
            )
        finally:
            tmux_manager.close_pane()

    def _wait_for_completion(self, tmux_manager: "TmuxManager") -> str:
        """Wait for bash command to finish using idle detection.

        Uses hash-based idle detection since bash commands don't use
        the Claude hook system.
        """
        start = time.time()
        waiter = AnimatedWaiter(tool_name="bash")

        # Hash-based idle detection state
        last_hash = ""
        last_hash_change_time = time.time()
        last_hash_check_time = 0.0
        hash_check_interval = 2.0  # Check more frequently for bash
        idle_timeout = 10.0  # Shorter timeout for bash commands

        with Live(console=get_display().console, refresh_per_second=10) as live:
            while True:
                elapsed = time.time() - start
                live.update(waiter.create_display(elapsed))

                # Hash-based idle detection
                current_time = time.time()
                if current_time - last_hash_check_time >= hash_check_interval:
                    last_hash_check_time = current_time
                    current_hash = tmux_manager.get_pane_content_hash()

                    if current_hash != last_hash:
                        # Content changed, reset timer
                        last_hash = current_hash
                        last_hash_change_time = current_time
                    elif current_time - last_hash_change_time >= idle_timeout:
                        # No change for idle_timeout seconds, consider done
                        break

                # Brief sleep before next iteration
                time.sleep(0.2)

        # Capture final output
        return tmux_manager.capture_pane_content()

File: src/core/tools/bash.ts
/**
 * Bash tool implementation.
 */

import type { ExecutionContext } from "../context/execution.ts";
import type { TmuxManager } from "../tmux/manager.ts";
import type { StepConfig } from "../../types/index.ts";
import { LoopSignal } from "../../types/index.ts";
import type { ToolResult } from "./types.ts";
import { BaseTool, successResult, errorResult } from "./types.ts";

/**
 * Execute bash commands in subprocess or tmux pane.
 */
export class BashTool extends BaseTool {
  get name(): string {
    return "bash";
  }

  validateStep(step: StepConfig): void {
    if (!step.command) {
      throw new Error("Bash step requires 'command' field");
    }
  }

  async execute(
    step: StepConfig,
    context: ExecutionContext,
    tmuxManager: TmuxManager
  ): Promise<ToolResult> {
    const command = context.interpolate(step.command!);
    const cwd = context.interpolateOptional(step.cwd) ?? context.projectPath;
    const visible = step.visible ?? false;
    const stripOutput = step.stripOutput ?? true;

    // Build environment variables if specified
    const env = this.buildEnv(step.env, context);

    if (visible) {
      return this.executeVisible(command, cwd, tmuxManager, stripOutput, env);
    }
    return this.executeSubprocess(command, cwd, stripOutput, env);
  }

  private buildEnv(
    envConfig: Record<string, string> | undefined,
    context: ExecutionContext
  ): Record<string, string> | undefined {
    if (!envConfig) {
      return undefined;
    }

    // Start with copy of current environment
    const env: Record<string, string> = { ...process.env } as Record<string, string>;

    // Add/override with custom variables (interpolated)
    for (const [key, value] of Object.entries(envConfig)) {
      env[key] = context.interpolate(String(value));
    }

    return env;
  }

  private async executeSubprocess(
    command: string,
    cwd: string,
    stripOutput: boolean,
    env?: Record<string, string>
  ): Promise<ToolResult> {
    console.log(`Running: ${command}`);

    try {
      const proc = Bun.spawn(["sh", "-c", command], {
        cwd,
        env,
        stdout: "pipe",
        stderr: "pipe",
      });

      // Set up timeout
      const timeout = 600_000; // 10 minutes
      const timeoutPromise = new Promise<never>((_, reject) => {
        setTimeout(() => {
          proc.kill();
          reject(new Error("Command timed out after 10 minutes"));
        }, timeout);
      });

      const [stdout, stderr] = await Promise.race([
        Promise.all([
          new Response(proc.stdout).text(),
          new Response(proc.stderr).text(),
        ]),
        timeoutPromise,
      ]);

      const exitCode = await proc.exited;

      let output = stdout || "";
      if (stderr) {
        output += `\n[STDERR]\n${stderr}`;
      }

      if (stripOutput) {
        output = output.trim();
      }

      const success = exitCode === 0;

      return {
        success,
        output,
        error: success ? undefined : stderr || undefined,
        loopSignal: LoopSignal.NONE,
      };
    } catch (error) {
      const message =
        error instanceof Error ? error.message : String(error);
      return errorResult(message);
    }
  }

  private async executeVisible(
    command: string,
    cwd: string,
    tmuxManager: TmuxManager,
    stripOutput: boolean,
    env?: Record<string, string>
  ): Promise<ToolResult> {
    // For visible mode with custom env vars, wrap the command with exports
    let finalCommand = command;
    if (env) {
      const exports: string[] = [];
      for (const [key, value] of Object.entries(env)) {
        // Skip system env vars that are already set with same value
        if (process.env[key] !== value) {
          const escapedValue = value.replace(/'/g, "'\\''");
          exports.push(`export ${key}='${escapedValue}'`);
        }
      }
      if (exports.length > 0) {
        finalCommand = `${exports.join(" && ")} && ${command}`;
      }
    }

    // Launch bash pane
    await tmuxManager.launchBashPane(finalCommand, cwd);

    try {
      // Wait for completion using idle detection
      const output = await this.waitForCompletion(tmuxManager);
      const finalOutput = stripOutput ? output.trim() : output;

      return successResult(finalOutput);
    } finally {
      await tmuxManager.closePane();
    }
  }

  private async waitForCompletion(tmuxManager: TmuxManager): Promise<string> {
    const startTime = Date.now();

    // Hash-based idle detection state
    let lastHash = "";
    let lastHashChangeTime = Date.now();
    let lastHashCheckTime = 0;
    const hashCheckInterval = 2000; // Check every 2 seconds
    const idleTimeout = 10000; // 10 seconds idle = done

    while (true) {
      const currentTime = Date.now();
      const elapsed = currentTime - startTime;

      // Update display periodically
      if (elapsed % 5000 < 200) {
        console.log(`Waiting... ${Math.floor(elapsed / 1000)}s`);
      }

      // Hash-based idle detection
      if (currentTime - lastHashCheckTime >= hashCheckInterval) {
        lastHashCheckTime = currentTime;
        const currentHash = await tmuxManager.getPaneContentHash();

        if (currentHash !== lastHash) {
          // Content changed, reset timer
          lastHash = currentHash;
          lastHashChangeTime = currentTime;
        } else if (currentTime - lastHashChangeTime >= idleTimeout) {
          // No change for idle timeout, consider done
          break;
        }
      }

      await Bun.sleep(200);
    }

    // Capture final output
    return tmuxManager.capturePaneContent();
  }
}
